integration-tests/src/test/java/com/datastax/oss/driver/mapper/QueryProviderIT.java

```diff
@@ -0,0 +1,224 @@
+import com.datastax.oss.driver.api.mapper.annotations.Entity;
+import com.datastax.oss.driver.api.mapper.annotations.Insert;
+import com.datastax.oss.driver.api.mapper.annotations.Mapper;
+import com.datastax.oss.driver.api.mapper.annotations.PartitionKey;
+import com.datastax.oss.driver.api.mapper.annotations.QueryProvider;
+import com.datastax.oss.driver.api.mapper.entity.EntityHelper;
+import com.datastax.oss.driver.api.querybuilder.select.Select;
+import com.datastax.oss.driver.api.testinfra.ccm.CcmRule;
+import com.datastax.oss.driver.api.testinfra.session.SessionRule;
+import com.datastax.oss.driver.categories.ParallelizableTests;
+import java.util.Objects;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+public class QueryProviderIT {
+
+  private static CcmRule ccm = CcmRule.getInstance();
+  private static SessionRule<CqlSession> sessionRule = SessionRule.builder(ccm).build();
+  @ClassRule public static TestRule chain = RuleChain.outerRule(ccm).around(sessionRule);
+
+  private static SensorDao dao;
+
+  @BeforeClass
+  public static void setup() {
+    CqlSession session = sessionRule.session();
+
+    session.execute(
+        SimpleStatement.builder(
+                "CREATE TABLE sensor_reading(sensor_id int, month int, day int, value double, "
+                    + "PRIMARY KEY (sensor_id, month, day)) "
+                    + "WITH CLUSTERING ORDER BY (month DESC, day DESC)")
+            .setExecutionProfile(sessionRule.slowProfile())
+            .build());
+
+    SensorMapper mapper = new QueryProviderIT_SensorMapperBuilder(session).build();
+    dao = mapper.sensorDao(sessionRule.keyspace());
+  }
+
+  @Test
+  public void should_invoke_query_provider() {
+    SensorReading readingFeb3 = new SensorReading(1, 2, 3, 9.3);
+    SensorReading readingFeb2 = new SensorReading(1, 2, 2, 8.6);
+    SensorReading readingFeb1 = new SensorReading(1, 2, 1, 8.7);
+    SensorReading readingJan31 = new SensorReading(1, 1, 31, 8.2);
+    dao.save(readingFeb3);
+    dao.save(readingFeb2);
+    dao.save(readingFeb1);
+    dao.save(readingJan31);
+
+    assertThat(dao.findSlice(1, null, null).all())
+        .containsExactly(readingFeb3, readingFeb2, readingFeb1, readingJan31);
+    assertThat(dao.findSlice(1, 2, null).all())
+        .containsExactly(readingFeb3, readingFeb2, readingFeb1);
+    assertThat(dao.findSlice(1, 2, 3).all()).containsExactly(readingFeb3);
+  }
+
+  @Mapper
+  public interface SensorMapper {
+    @DaoFactory
+    SensorDao sensorDao(@DaoKeyspace CqlIdentifier keyspace);
+  }
+
+  @Dao
+  public interface SensorDao {
+    @QueryProvider(providerClass = FindSliceProvider.class, entityHelpers = SensorReading.class)
+    PagingIterable<SensorReading> findSlice(int id, Integer month, Integer day);
```

Comment:
It took me a little bit of time to understand the value of this, but now i'm warming up to it :).  This feels like a 'power user' thing, where I expect most to be comfortable with existing capabilities and not need this level of flexibility.

I think the main things are:

1. Provides a level of customization at Dao level that didn't previously exist without injecting statement/statementbuilder or ResultSet/Row into the dao interface.  If you wanted to do something fancy previously, you'd either have to use `@Query` to provide CQL, pass in a `BoundStatementBuilder` to inject entity values or on the result mapping side, or pass in a ResultSet/Row from an executed query to map the result back.

2. Having access to `EntityHelper`s gives you direct access to methods for mapping between entities and results, getting statement basis and so forth.

Am I understanding that right?

Reply:
Yes, definitely power user.

1. The main point for me is the ability to build the query dynamically depending on the parameters. With all the other annotations the query string is completely defined at compile time, so yeah the only alternative would be to execute the query outside of the mapper, and only use `GetEntity` / `SetEntity` to inject your parameters and extract the results. This is essentially that, but more cleanly encapsulated, from the caller's POV it's just another DAO method.

2. Correct, entity helpers give you access to the lower-level building blocks used internally by the mapper.